File: docs/current_docs/manuals/developer/go/snippets/services/service-lifecycle-1/main.go
package main

import (
	"context"
)

type MyModule struct{}

func (m *MyModule) RedisService(ctx context.Context) (string, error) {
	// create Redis service container
	redisSrv := dag.Container().
		From("redis").
		WithExposedPort(6379).
		AsService()

	// create Redis client container
	redisCLI := dag.Container().
		From("redis").
		WithServiceBinding("redis-srv", redisSrv).
		WithEntrypoint([]string{"redis-cli", "-h", "redis-srv"})

	// send ping from client to server
	ping := redisCLI.WithExec([]string{"ping"})
	return ping.Stdout(ctx)
}

File: docs/current_docs/manuals/developer/snippets/services/service-lifecycle-1/python/main.py
from dagger import dag, function, object_type


@object_type
class MyModule:
    @function
    async def redis_service(self) -> str:
        """Creates Redis service and client."""
        redis_srv = dag.container().from_("redis").with_exposed_port(6379).as_service()

        # create Redis client container
        redis_cli = (
            dag.container()
            .from_("redis")
            .with_service_binding("redis-srv", redis_srv)
            .with_entrypoint(["redis-cli", "-h", "redis-srv"])
        )

        # send ping from client to server
        return await redis_cli.with_exec(["ping"]).stdout()
